BUG: Fix Series.reindex losing values when reindexing to MultiIndex

Roline-Stapny opened the pull request:
- [X] closes #60923
- [X ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [X] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [X] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [X] Added an entry in the latest `doc/source/whatsnew/v3.0.0.rst` file if fixing a bug or adding a new feature.

##Series.reindex()

#Before
```
# Create a Series with a named Index
series = pd.Series([26.73, 24.255], index=pd.Index([81, 82], name='a'))

# Create a MultiIndex with level names 'a', 'b', 'c'
target = pd.MultiIndex.from_product(
    [[81, 82], [np.nan], ["2018-06-01", "2018-07-01"]], 
    names=["a", "b", "c"]
)

# This would incorrectly set all values to NaN
series.reindex(target)
# a   b    c         
# 81  NaN  2018-06-01   NaN
#          2018-07-01   NaN
# 82  NaN  2018-06-01   NaN
#          2018-07-01   NaN

# But this works correctly
series.reindex(target, level="a")
# a   b    c         
# 81  NaN  2018-06-01    26.73
#          2018-07-01    26.73
# 82  NaN  2018-06-01    24.255
#          2018-07-01    24.255

```

#After
```
# Same setup as before
series = pd.Series([26.73, 24.255], index=pd.Index([81, 82], name='a'))
target = pd.MultiIndex.from_product(
    [[81, 82], [np.nan], ["2018-06-01", "2018-07-01"]], 
    names=["a", "b", "c"]
)

# Now both produce the same correct result
series.reindex(target)  # Automatically detects level='a'
# a   b    c         
# 81  NaN  2018-06-01    26.73
#          2018-07-01    26.73
# 82  NaN  2018-06-01    24.255
#          2018-07-01    24.255

```

##Datafram.reindex()



```
df = pd.DataFrame({
    'value': [26.73, 24.255],
    'other': ['A', 'B']
}, index=pd.Index([81, 82], name='a'))

target = pd.MultiIndex.from_product(
    [[81, 82], [np.nan], ["2018-06-01", "2018-07-01"]], 
    names=["a", "b", "c"]
)
```

Before
```
df.reindex(index = target)
                   value other
a  b   c
81 NaN 2018-06-01    NaN   NaN
       2018-07-01    NaN   NaN
82 NaN 2018-06-01    NaN   NaN
       2018-07-01    NaN   NaN
```

After

```
df.reindex(index = target)
                    value other
a  b   c
81 NaN 2018-06-01  26.730     A
       2018-07-01  26.730     A
82 NaN 2018-06-01  24.255     B
       2018-07-01  24.255     B
```

Commit:
BUG: Fix Series.reindex losing values when reindexing to MultiIndex

mroeschke commented on pandas/tests/series/methods/test_reindex.py:
```diff
@@ -434,3 +434,49 @@ def test_reindex_expand_nonnano_nat(dtype):
+    "idx,expected_match_level_a",
+    [
+        # Source index has matching name - should match level "a"
+        (Index([81, 82], name="a"), True),
```
Instead of `idx` could you parameterize on `name` instead since the data is the same?

Roline-Stapny replied:
yes, Ill update in the next revision

mroeschke reviewed (changes requested):
Does `DataFrame.reindex` also need the same handling?

Commit:
Merge branch 'main' into sroline_issue_60923

Commit:
BUG: Fix dataframe.reindex to detect matching level when source has single Index

mroeschke commented on doc/source/whatsnew/v3.0.0.rst:
```diff
@@ -788,6 +788,8 @@ MultiIndex
 - Bug in :class:`DataFrame` arithmetic operations in case of unaligned MultiIndex columns (:issue:`60498`)
 - Bug in :class:`DataFrame` arithmetic operations with :class:`Series` in case of unaligned MultiIndex (:issue:`61009`)
 - Bug in :meth:`MultiIndex.from_tuples` causing wrong output with input of type tuples having NaN values (:issue:`60695`, :issue:`60988`)
+- Bug in :meth:`DataFrame.reindex` where reindexing a Dataframe with a named Index to a MultiIndex would incorrectly set all values to ``NaN``. Now correctly preserves values when the source index name matches a target level name (:issue:`60923`)
+- Bug in :meth:`Series.reindex` where reindexing a Series with a named Index to a MultiIndex would incorrectly set all values to ``NaN``. Now correctly preserves values when the source index name matches a target level name (:issue:`60923`)
```
```suggestion
- Bug in :meth:`DataFrame.reindex` and :meth:`Series.reindex` where reindexing :class:`Index` to a :class:`MultiIndex` would incorrectly set all values to ``NaN``.(:issue:`60923`)
```

mroeschke commented on pandas/core/frame.py:
```diff
@@ -5355,6 +5355,18 @@ def reindex(
+        # This prevents values from being incorrectly set to NaN when the source index
+        # name matches a level name in the target MultiIndex. Only applies when source
+        # is not already a MultiIndex.
+        if (
+            level is None
+            and index is not None
+            and isinstance(index, MultiIndex)
+            and not isinstance(self.index, MultiIndex)
+            and self.index.name in index.names
+        ):
+            level = self.index.name
```
Is it possible to combine this logic with the `series.py` logic and push it into `generic.py`?

mroeschke commented on pandas/tests/frame/methods/test_reindex.py:
```diff
@@ -1258,3 +1258,55 @@ def test_invalid_method(self):
             df.reindex([1, 0, 2], method="asfreq")
+
+    def test_reindex_index_name_matches_multiindex_level(self):
+        """
+        Test automatic level detection when reindexing from Index to MultiIndex.
+        When source index name matches a level name in target MultiIndex and level
+        is not specified, should behave same as if level was explicitly set.
+        """
+        # Create source DataFrame with named Index
```
Please remove all test docstrings and comments

mroeschke commented on pandas/tests/frame/methods/test_reindex.py:
```diff
@@ -1258,3 +1258,55 @@ def test_invalid_method(self):
+
+        result = df.reindex(index=target)
+
+        # Should fill with NaN since no level match
+        assert result.isna().all().all()
+
+        # Verify shape is correct
+        assert result.index.equals(target)
+        assert result.columns.equals(df.columns)
```
Create an `expected = ` result to compare to

Commit:
Merge remote-tracking branch 'upstream/main' into sroline_issue_60923

Commit:
BUG: Fix DataFrame.reindex and Series.reindex loosing values when reindexing single to multiindex

Commit:
Merge remote-tracking branch 'upstream/main' into sroline_issue_60923

Commit:
Automatic level detection for dataframe reindex moved to generic.py and remove non-relevant test

Roline-Stapny commented:
@mroeschke , sorry for tagging again. I have addressed all comments from what I can see,  but still seeing "Changes Requested" "[mroeschke](https://github.com/mroeschke) Requested changes". Have I missed addressing any of your comments? Do you mind pointing which comment is not being addressed.

mroeschke commented on pandas/tests/frame/methods/test_reindex.py:
```diff
@@ -1258,3 +1258,28 @@ def test_invalid_method(self):
+        )
+
+        result = df.reindex(index=target)
+        expected = df.reindex(index=target, level="a")
```
Could you build `expected` from a new DataFrame, e.g. `expected = DataFrame(data=..., index=..., ...)` instead of using `reindex`?

mroeschke reviewed (commented):
Just a few comments otherwise looks good

Commit:
Merge remote-tracking branch 'upstream/main' into sroline_issue_60923

Commit:
BUG: Fix Series.reindex losing values when reindexing to MultiIndex - build test expected results

mroeschke commented:
Thanks @Roline-Stapny
